hedera-node/hedera-network-service-impl/src/main/java/com/hedera/node/app/service/network/impl/ReadableRunningHashLeafStoreImpl.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.network.impl;
+
+import static com.hedera.node.app.service.network.impl.NetworkServiceImpl.RUNNING_HASHES_KEY;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.node.app.service.mono.stream.RecordsRunningHashLeaf;
+import com.hedera.node.app.service.network.ReadableRunningHashLeafStore;
+import com.hedera.node.app.spi.state.ReadableSingletonState;
+import com.hedera.node.app.spi.state.ReadableStates;
+import com.swirlds.common.crypto.RunningHash;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * The underlying data storage class that holds the record running hash leaf data.
+ */
+public class ReadableRunningHashLeafStoreImpl implements ReadableRunningHashLeafStore {
```

Comment:
I believe the app will need to expose the `n-3` running hash through a custom interface, since the app will be managing the record stream in state as a special Merkle leaf outside the `States` API.

So although this is quite reasonable for this PR, I would slightly prefer---as a first approximation---to add a `RunningHash getTrailingHash(int n)` method on the `HandleContext`.

(Note the `ContractService` will **also** need the `n-1` trailing hash, both for its independent cache of block hashes and to implement `blockhash(block.number - 1)` in a transaction that is first in a block.)

Reply:
Wouldn't NetworkService be responsible for the record running hash leaf ?